coordinator/src/main/scala/com/socrata/datacoordinator/service/collocation/Collocator.scala

```diff
@@ -0,0 +1,275 @@
+
+object CollocationResult {
+  implicit val encode = AutomaticJsonEncodeBuilder[CollocationResult]
+
+  def canonicalEmpty: CollocationResult = CollocationResult(
+    status = Completed,
+    message = Completed.message,
+    cost = Cost(0),
+    moves = Seq.empty
+  )
+}
+
+trait Collocator {
+  def collocatedDatasets(datasets: Set[DatasetInternalName]): Either[RequestError, CollocatedDatasetsResult]
+  def explainCollocation(secondaryGroup: String,
+                         request: CollocationRequest): Either[RequestError, Either[ResourceNotFound, CollocationResult]]
+  def executeCollocation(secondaryGroup: String,
+                         request: CollocationRequest): Either[RequestError, Either[ResourceNotFound, CollocationResult]]
+}
+
+trait CollocatorProvider {
+  val collocator: Collocator
+}
+
+class CoordinatedCollocator(collocationGroup: Set[String],
+                            coordinator: Coordinator,
+                            lock: CollocationLock,
+                            lockTimeoutMillis: Long) extends Collocator {
+
+  private val log = org.slf4j.LoggerFactory.getLogger(classOf[Collocator])
+
+  override def collocatedDatasets(datasets: Set[DatasetInternalName]): Either[RequestError, CollocatedDatasetsResult]  = {
+    try {
+      @tailrec
+      def search(current: Set[DatasetInternalName], seen: Set[DatasetInternalName]): Set[DatasetInternalName] = {
+        val toExplore = collocationGroup.flatMap { instance =>
+          coordinator.collocatedDatasetsOnInstance(instance, current).fold(throw _, _.datasets)
+        } -- seen
+
+        if (toExplore.nonEmpty) search(toExplore, seen ++ toExplore)
+        else seen
+      }
+
+      Right(CollocatedDatasetsResult(search(current = datasets, seen = datasets)))
+    } catch {
+      case error: RequestError => Left(error)
+    }
+  }
+
+  override def explainCollocation(secondaryGroup: String, request: CollocationRequest):
+  Either[RequestError, Either[ResourceNotFound, CollocationResult]] = {
+    try {
+      lock.acquire(lockTimeoutMillis) match {
+        case Some(timeout) =>
+          log.warn("Timeout while waiting to acquire collocation lock to execute collocation.")
+          throw timeout
+        case None =>
+          coordinator.secondaryGroups.get(secondaryGroup) match {
+            case Some(groupConfig) =>
+              val instances = groupConfig.instances
+              val unavailableInstances = groupConfig.instancesNotAcceptingNewDatasets.getOrElse(Set.empty) // TODO: use in logic
+            val replicationFactor = groupConfig.numReplicas
+
+              request match {
+                case CollocationRequest(collocations) if collocations.isEmpty => Right(Right(CollocationResult.canonicalEmpty))
+                case CollocationRequest(collocations) =>
+                  // TODO: verify collocations (datasets exist and columns exist with matching types)
+
+                  def partitionedNodes[T](edges: Set[Set[T]]): Set[Set[T]] = {
```

Comment:
These functions should probably be pulled out somewhere with some comments and unit testing and to 
split the generic "this is a graph utility function" code from "this is the real logic of what we are trying to do" so it is obvious at first glance and so they can be reasoned about separately.  This sort of logic seems conducive to unit testing, which would also help make it easier to review by providing examples.